c2rust-analyze/src/main.rs

```diff
@@ -291,9 +293,14 @@ fn run(tcx: TyCtxt) {
     }
     eprintln!("reached fixpoint in {} iterations", loop_count);
 
-    // Print results for each function.
+    // Print results for each function.  We use declaration order (as reported by `body_owners`)
+    // since this makes FileCheck tests easier to write.
     for ldid in tcx.hir().body_owners() {
-        let info = func_info.get_mut(&ldid).unwrap();
+        // Skip any body owners that aren't listed in `func_info`.
+        let info = match func_info.get_mut(&ldid) {
+            Some(x) => x,
+            None => continue,
```

Comment:
does this avoid foreign functions declared with `extern "C"` for example?

Reply:
Functions declared in `extern` blocks aren't body owners, so this loop won't visit them in the first place.

I updated the comment on the loop header as follows:
```Rust
    // Print results for each function in `all_fn_ldids`, going in declaration order.  Concretely,
    // we iterate over `body_owners()`, which is a superset of `all_fn_ldids`, and filter based on
    // membership in `func_info`, which contains an entry for each ID in `all_fn_ldids`.
```
I think this better explains the logic here - let me know if it's still unclear.

Reply:
It looks like those functions count as body owners somehow, or something odd is happening. If `calloc` is declared `extern "C"` in a contrived lighttpd example:

```
error: internal compiler error: compiler/rustc_typeck/src/check/mod.rs:364:9: can't type-check body of DefId(0:13 ~ lighttpd[c91b]::{extern#0}::calloc)
  --> ../analysis/test/src/lighttpd.rs:11:5
   |
11 |     fn calloc(_: libc::c_ulong, _: libc::c_ulong) -> *mut libc::c_void;
   |
```

which happens in the `mir_built` call in `for_each_callee`

Reply:
ah, I see the problem, we don't filter by body owners when they are callees:

```rust
if let Some(Callee::Other { def_id, .. }) = util::ty_callee(self.tcx, ty) {
  if let Some(ldid) = def_id.as_local() {
    (self.f)(ldid);
  }
}
```

Reply:
another follow-up, it would appear that `as_local` should return `None` for something declared `extern "C"` but for the `calloc` example it returns `Some(DefId(0:13 ~ lighttpd[c91b]::{extern#0}::calloc))`

Reply:
The `as_local` behavior here is correct.  The Rust declaration of `calloc` as an extern fn is local to the crate, even if the actual code for that function is not.

I'll add a test case using extern functions and investigate further what's going wrong here.

Reply:
Added a test case that shows the boundary of what's currently supported for foreign fns.  I fixed the immediate panic, so you can mention a foreign function (e.g. `let f = malloc;`), but you still can't call them, the same way you can't call unsupported library functions.